lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py

```diff
@@ -0,0 +1,283 @@
+
+- name: Delete a dynamic resource definition in ManageIQ
+  manageiq_dynamic_resource_definition:
+    name: 'my_edited_custom_definition'
+    state: 'absent'
+    manageiq_connection:
+      url: 'http://127.0.0.1:3000'
+      username: 'admin'
+      password: 'smartvm'
+      verify_ssl: False
+'''
+
+RETURN = '''
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.manageiq import ManageIQ, manageiq_argument_spec
+
+
+class ManageIQDynamicResourceDefinition(object):
+    """
+        Object to create dynamic resource definition operations in manageiq.
+    """
+
+    def __init__(self, manageiq):
+        self.manageiq = manageiq
+        self.module = self.manageiq.module
+        self.api_url = self.manageiq.api_url
+        self.client = self.manageiq.client
+
+    def dynamic_resource_definition(self, name):
+        """ Search for dynamic resource definition by name name.
+
+        Returns:
+            the dynamic resource definition, or None if not found.
+        """
+        return self.manageiq.find_collection_resource_by('generic_object_definitions', name=name)
+
+    def create_definition(self, dynamic_resource_definition, name, properties):
+        """ Creates the dynamic resource definition in manageiq.
+
+        Returns:
+            a dictionary with changed and msg.
+        """
+        url = '%s/generic_object_definitions' % (self.api_url)
+
+        resource = {
+            'name': name,
+            'properties': properties,
+        }
+
+        try:
+            result = self.client.post(url, action='create', resource=resource)
+        except Exception as e:
+            self.module.fail_json(
+                msg="failed to create dynamic resource definition %s: %s" % (name, str(e))
+            )
+
+        return dict(
+            changed=True,
+            msg="successfully created the dynamic resource definition %s: %s" % (name, result['results'])
+        )
+
+    def has_field(self, dynamic_resource_definition, key):
+        """ Check if definition properties has a specific key.
+
+        If key is unhashble, return.
+        """
+        try:
+            if dynamic_resource_definition and dynamic_resource_definition.get('properties', {}).get(key):
+                return True
+        except TypeError:
+            return
+        else:
+            return
+
+    def edit_definition(self, dynamic_resource_definition, name, properties):
+        """ Edit the dynamic resource definition in manageiq.
+
+        Returns:
+            a dictionary with the changed and msg.
+        """
+        # Check if properties has the following keys,
+        # These keys are returned from the API response even if they are empty
+        # So we remove them to compare accuratley
+        for key in ['attributes', 'associations', 'methods']:
+            if not self.has_field(dynamic_resource_definition, key):
+                dynamic_resource_definition['properties'].pop(key)
```

Comment:
OK, I see what you're doing here :+1:, I think there is slightly better way.  Just suggestions, your call.

Just removing these is not perfect.  Ideally, omitting say `associations` in module args would behave same as writing explicit `associations: {}`.  But currently explicit `associations: {}` won't match what's left after you remove empty values from API, causing unnecessary request and changed=True.
- Args side can be normalized for you if you make attributes & associations first-class options, or suboptions, defaulting to `{}`.
  - Note that only recently manageiq master started sending empty values; nulls were omitted before.  Not sure that's relevant — do you get Nones or empty hashes here?  Anyway :+1: to normalizing API results in one direction or another — don't *rely* on getting empty values.
- A pattern that works & reads nicely is defining a normalizing function and applying it on both sides of comparison.  `if normalize(current) == normalized(desired)`.
- IMHO, splitting logic between `has_field` and here is harder to follow, and it could be simple enough to inline, perhaps something like this:
  ```
  current_properties = dynamic_resource_definition.get('properties', {})
  current_name = dynamic_resource_definition.get('name')
  if (current_name == name and
      current_properties.get('attributes', {}) == properties.get('attributes', {}) and
      current_properties.get('associations', {}) == properties.get('associations', {}):
  ```

Reply:
I see what you mean. I agree, the inline solution is more simple and straight forward